query/src/main/java/org/infinispan/query/clustered/DistributedEntryIterator.java

```diff
@@ -16,14 +22,41 @@
  */
 class DistributedEntryIterator<K, V> extends DistributedIterator<EntityEntry<K, V>> {
 
+   private final boolean withMetadata;
+   private final QueryKeyConverter queryKeyConverter;
+
    DistributedEntryIterator(LocalQueryStatistics queryStatistics, Sort sort, int resultSize,
                             int maxResults, int firstResult, Map<Address, NodeTopDocs> topDocsResponses,
-                            AdvancedCache<?, ?> cache) {
+                            AdvancedCache<?, ?> cache, boolean withMetadata) {
       super(queryStatistics, sort, resultSize, maxResults, firstResult, topDocsResponses, cache);
+      this.withMetadata = withMetadata;
+      queryKeyConverter = new QueryKeyConverter(cache);
    }
 
    @Override
-   protected EntityEntry<K, V> decorate(Object key, Object value, float score) {
-      return new EntityEntry<>((K) key, (V) value, score);
+   protected EntityEntry<K, V> decorate(Object key, Object value, float score, Metadata metadata) {
+      return new EntityEntry<>((K) key, (V) value, score, metadata);
+   }
+
+   @Override
+   protected void getAllAndStore(List<KeyAndScore> keysAndScores) {
+      if (!withMetadata) {
+         super.getAllAndStore(keysAndScores);
+         return;
+      }
+
+      Set<Object> keySet = keysAndScores.stream()
+            .map(keyAndScore -> keyAndScore.key)
+            .collect(Collectors.toSet());
+      Map<?, CacheEntry<Object, Object>> convertedMap =
+            queryKeyConverter.convertEntries(cache.getAllCacheEntries(keySet));
+      keysAndScores.forEach(bla -> bla.covertedKey = queryKeyConverter.covert(bla.key));
```

Comment:
maybe it is because it is late, but I don't understand why do you need to convert. Can you explain? 
It seems to convert the key but it does nothing with it.

Reply:
The problem is that we have a `Map<byte[], Object>` here. If we use the `cache.getAll()` this is not a problem, since we have the memory items. If we use `cache.getAllCacheEntries()` the `byte[]` returned by API are not the same object, so I need to convert them. If you try to run the tests without the conversion, you can verify the issue.

Reply:
I think this can be done without converting, let me take a closer look. This code should be able to be a lot simpler.

Reply:
I haven't found any other solution to this issue. Of course if we could simplify it, I would be very happy. Thanks for looking at it @wburns

Reply:
You can do something like this:

```java
      Map<Object, Float> keyScores = keysAndScores.stream()
            .collect(Collectors.toMap(kas ->
                  // Have to convert in case the storage is a byte[] which doesn't work with Map lookups
               queryKeyConverter.covert(kas.key)
            , KeyAndScore::getScore));

      StreamSupport.stream(cache.cacheEntrySet().stream().filterKeys(keyScores.keySet())
                  .spliterator(), false)
            .map(e -> decorate(e.getKey(), e.getValue(), keyScores.get(queryKeyConverter.covert(e.getKey())), e.getMetadata()))
            .forEach(values::add);
```

Note that this avoids sending the map function and returns marshalled CacheEntry instances.

If it is acceptable to serialize the EntityEntry instance instead you can remove the whole StreamSupport part and do

```java
      Map<Object, Float> keyScores = keysAndScores.stream()
            .collect(Collectors.toMap(kas ->
                  // Have to convert in case the storage is a byte[] which doesn't work with Map lookups
               queryKeyConverter.covert(kas.key)
            , KeyAndScore::getScore));

      cache.cacheEntrySet().stream().filterKeys(keyScores.keySet())
            .map(e -> decorate(e.getKey(), e.getValue(), keyScores.get(queryKeyConverter.covert(e.getKey())), e.getMetadata()))
            .iterator().forEachRemaining(values::add);
```

Reply:
Note above the spliterator and iterator methods are the terminal operations for the stream. So any operations before them would be done remotely if this is a remote cache.

Reply:
Also a side note, you should pretty much always prefer `cache.entrySet.stream` over `getAll` as it allow for streaming of results and remote execution of intermediate operations.

Reply:
I'm not sure if we can use stream if the order of the results matters. 
If we can use stream, it would be better to re-write the `DistributedIterator` to iterate over Lucene results to fetch the keys on demand. 
`getAllAndStore()` is to fetch the entries in batch instead of having a huge `getAll`

I need to attach a debugger to check this method. I'm surprised `getAll()` does not need to convert keys while this one does...

Reply:
Ah, true I forgot about the order. Then we can probably axe this idea then.

Reply:
good catch @pruivo! Yes we really need to preserve the order here. I'm resolving this conversation then.

Reply:
You can eliminate the conversion if you use the original key in the `EncoderCache`. I tested the following changes:

```diff
--- a/core/src/main/java/org/infinispan/cache/impl/EncoderCache.java
+++ b/core/src/main/java/org/infinispan/cache/impl/EncoderCache.java
@@ -150,13 +150,12 @@ private CacheEntry<K, V> convertEntry(K newKey, V newValue, CacheEntry<K, V> ent
 
    private Map<K, CacheEntry<K, V>> decodeEntryMapForRead(Map<K, CacheEntry<K, V>> map) {
       Map<K, CacheEntry<K, V>> entryMap = new HashMap<>(map.size());
-      map.values().forEach(v -> {
-         K originalKey = v.getKey();
-         K unwrappedKey = keyFromStorage(originalKey);
+      map.forEach((k, v) -> {
+         K unwrappedKey = keyFromStorage(k);
          V originalValue = v.getValue();
          V unwrappedValue = valueFromStorage(originalValue);
          CacheEntry<K, V> entryToPut;
-         if (unwrappedKey != originalKey || unwrappedValue != originalValue) {
+         if (unwrappedKey != k || unwrappedValue != originalValue) {
             entryToPut = convertEntry(unwrappedKey, unwrappedValue, v);
          } else {
             entryToPut = v;

```

Reply:
Sure. But I would prefer to not touch the `EncoderCache` in the context of this contribution. Feel free to open a followup to modify the `EncoderCache` and remove the conversion.

Reply:
Fair enough.